apis/supervisor/config/v1alpha1/types_federationdomain.go.tmpl

```diff
@@ -42,6 +42,155 @@ type FederationDomainTLSSpec struct {
 	SecretName string `json:"secretName,omitempty"`
 }
 
+// FederationDomainTransformsConstant defines a constant variable and its value which will be made available to
+// the transform expressions. This is a union type, and Type is the discriminator field.
+type FederationDomainTransformsConstant struct {
+	// Name determines the name of the constant. It must be a valid identifier name.
+	// +kubebuilder:validation:Pattern=`^[a-zA-Z][_a-zA-Z0-9]*$`
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=64
+	Name string `json:"name"`
+
+	// Type determines the type of the constant, and indicates which other field should be non-empty.
+	// +kubebuilder:validation:Enum=string;stringList
+	Type string `json:"type"`
+
+	// StringValue should hold the value when Type is "string", and is otherwise ignored.
+	// +optional
+	StringValue string `json:"stringValue,omitempty"`
+
+	// StringListValue should hold the value when Type is "stringList", and is otherwise ignored.
+	// +optional
+	StringListValue []string `json:"stringListValue,omitempty"`
+}
+
+// FederationDomainTransformsExpression defines a transform expression.
+type FederationDomainTransformsExpression struct {
+	// Type determines the type of the expression. It must be one of the supported types.
+	// +kubebuilder:validation:Enum=policy/v1;username/v1;groups/v1
+	Type string `json:"type"`
+
+	// Expression is a CEL expression that will be evaluated based on the Type during an authentication.
+	// +kubebuilder:validation:MinLength=1
+	Expression string `json:"expression"`
+
+	// Message is only used when Type is policy/v1. It defines an error message to be used when the policy rejects
+	// an authentication attempt. When empty, a default message will be used.
+	// +optional
+	Message string `json:"message,omitempty"`
+}
+
+// FederationDomainTransformsExample defines a transform example.
+type FederationDomainTransformsExample struct {
```

Comment:
Not sure I'm sold on `Example` over `Test` or `Validation`.  I think the concept of example is more like documentation, but since the series of "example/expect" here has the power to shut down the FederationDomain if invalid I'm leaning toward a more assertive name.

Reply:
How about `expressionValidators`?

Reply:
`expressionValidator` ties it to `expression`, which over time might be helpful if we put other things on a federation domain.  `example` alone might be confusing "example o what?" if there are a number of things in the future.

Reply:
Considering the following:
```
transforms
  constants
  expressions
  # but I still prefer something more assertive
  validators
  expressionValidators
  examples
```

Reply:
I don't think there's a perfect name here. If we wanted to use something like `expressionValidators` then it would probably need to be the more awkward double plural `expressionsValidators` since there are multiple validators which each validate multiple expressions. The Kubernetes API uses the word `validations` to mean things that act more like our policy expressions because they run repeatedly in response to some action that happens on the cluster. These `examples` happen at create/update time for CR, so it would be nice to use a different word for them.